### ISSUE TITLE
http/2: Trailers

### ISSUE THREAD

**Issue by shirhatti** [aspnet/KestrelHttpServer#2630]
We need to re-visit trailers to design a consistent consumption model whether you're using http/1.1 or http/2

**Comment by Tratcher** [comment]
Duplicate:
https://github.com/aspnet/KestrelHttpServer/issues/2051
https://github.com/aspnet/KestrelHttpServer/issues/622

**Comment by mkosieradzki** [comment]
Could you consider bumping the priority for this feature, please? It is a "blocker" for https://github.com/grpc/grpc/issues/15139#issuecomment-406907161

/cc: @davidfowl

**Comment by mkosieradzki** [comment]
`Http2FrameWriter` always sends `END_STREAM` flag in the last `DATA_FRAME`. This behavior does not leave any place for sending trailers which should be sent in a separate `HEADERS_FRAME` with `END_STREAM` flag after the last `DATA_FRAME`.

/cc: @davidfowl

**Comment by Tratcher** [comment]
Yes this is still being considered for 2.2, but we have a few critical items ahead of it (like proper flow control).

**Comment by mkosieradzki** [comment]
@Tratcher I can try to create a PR to fix this or are you not accepting PRs to this repo?

**Comment by Tratcher** [comment]
We are, but this one is somewhat complex so we want to make sure we have a solid design first. The proposals so far are incomplete. I anticipate working on it in a few weeks.

**Comment by mkosieradzki** [comment]
@Tratcher - Sounds reasonable. Just letting you know that the lack of trailers support is a blocker for creating a gRPC middleware (https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md), because gRPC uses trailers to send the grpc-status.

**Comment by Tratcher** [comment]
@mkosieradzki do you primarily need to send trailers, receive them, or both? And is your need specific to HTTP/2 or would you also need it in HTTP/1.1? That matrix makes for for distinct implementations and we want to make sure we're focusing on the right ones.

**Comment by mkosieradzki** [comment]
@Tratcher HTTP/2 only, I can use a specific feature for this purpose or even a pub-ternal Http2Stream - I really need to send trailers, otherwise there is no way to send the gRPC status. I will check if there is a scenario when I need to receive trailers, but I don't think so.

For purpose of the gRPC-Web (HTTP/1.1 and web browser compatible) trailers are sent as part of the body, so those are no real trailers.

**Comment by Tratcher** [comment]
Ok, check out my design notes on https://github.com/aspnet/KestrelHttpServer/issues/622#issuecomment-411880791

**Comment by mkosieradzki** [comment]
Thanks, I will have a look. Here is the exact spec I am implementing: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md

**Comment by muratg** [comment]
cc @blowdart since we briefly talked about this yesterday. Any concerns with this?

**Comment by blowdart** [comment]
The general concern is around badly behaved reverse proxies which sees one of the host header, and the web app sees another.

So if this is only for outbound then it's not out problem. For inbound it becomes more interesting.

**Comment by Tratcher** [comment]
The trailers arrive after the request body so a reverse proxy would have to fully buffer the content to even begin to mix them up, and then it would likely merge them all into request headers rather than to continue to send them as trailers. We plan to keep the inbound trailers in a separate collection from the normal headers to limit that kind of confusion.

**Comment by Tratcher** [comment]
In 2.2.0-preview3 we've completed the work for HTTP/2 response trailers. Request trailers we've punted to a future release.
